### PR TITLE
ULMS-325 Added clients for Event and Conference

### PR DESCRIPTION
**Pull request by alexkonst**

### CODE REVIEW

File: src/event.js [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+
+  /**
+   * Read state
+   * @param roomId
+   * @param {String[]} sets
+   * @param {Object} filterParams
+   * @returns {Promise}
+   */
+  readState (roomId, sets, filterParams = {}) {
+    const {
```
Comment (rogaldh) [inline src/event.js]: Validation

File: src/event.js [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+   * @param roomId
+   * @param {Object} filterParams
+   * @returns {Promise}
+   */
+  listAgent (roomId, filterParams = {}) {
+    const { limit, offset } = filterParams
+    const params = {
+      limit,
+      offset,
+      room_id: roomId
```
[the remark this answers is not in the document]
Comment (rogaldh) [reply src/event.js]: I mean that there is no need to rename as the argument is already here

File: src/mqtt-rpc.js [hunk 3]
```diff
@@ -0,0 +1,151 @@
+import { v4 as uuid4 } from 'uuid'
+
+class MQTTRPCService {
+  constructor (mqttClient, topicIn, topicOut, codec, methods) {
+    this._codec = codec
+    this._handlerMap = {}
+    this._incomingRequestMap = {}
+    this._mqtt = mqttClient
```
Comment (rogaldh) [inline src/mqtt-rpc.js]: I prefer a separate method/setter for the reference-like entities. So we do not mix configuration and dependencies.

File: src/mqtt-rpc.js [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    this._methods = methods
+    this._requestStorage = {}
+    this._topicIn = topicIn
+    this._topicOut = topicOut
+
+    this._addSubscription()
```
Comment (rogaldh) [inline src/mqtt-rpc.js]: Is it a good idea to make a subscription on instantination?

File: src/mqtt-rpc.js [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        this._processResponse('resolve', correlationData, payload)
+      } else {
+        this._processResponse('reject', correlationData, payload)
+      }
+    } else if (type === 'request' && method && correlationData) {
+      this._processIncomingRequest(properties, payload)
+    } else if (type === 'event' && typeof label === 'string') {
+      this._processNotification(label, payload)
+    } else {
+      console.debug('Unknown message', message)
```
Comment (rogaldh) [inline src/mqtt-rpc.js]: Let's not use the console as debugger at the client.

File: src/mqtt-rpc.js [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+    }
+  }
+
+  _processIncomingRequest (properties, payload) {
+    const { correlationData, responseTopic, userProperties: { method } } = properties
+    const handler = this._handlerMap[method]
+    let result = null
```
Comment (rogaldh) [inline src/mqtt-rpc.js]: What's the point to use `let` as it's used once?

File: src/mqtt-rpc.js [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+      const properties = {
+        correlationData,
+        userProperties: {
+          local_timestamp: Date.now().toString(),
+          status: '200',
+          type: 'response'
+        }
+      }
+
+      this._mqtt.publish(responseTopic, this._codec.encode(result), { properties })
```
Comment (rogaldh) [inline src/mqtt-rpc.js]: Maybe we should handle errors as described at `_processRequest`

File: src/mqtt-rpc.js [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+    if (request) {
+      if (action === 'resolve') {
+        request.resolve(response)
+      } else {
+        request.reject(response)
+      }
+
+      delete this._requestStorage[id]
+    } else {
+      console.debug('Unknown response', id, response)
```
Comment (rogaldh) [inline src/mqtt-rpc.js]: Let's not use the `console`.

File: src/mqtt-rpc.js [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+    } else {
+      throw new Error(`Method ${method} already registered`)
+    }
+  }
+
+  unregister (method) {
+    if (this._handlerMap[method]) {
+      delete this._handlerMap[method]
+    } else {
+      throw new Error(`Method ${method} not registered`)
```
Comment (rogaldh) [inline src/mqtt-rpc.js]: was not ...

File: src/mqtt-rpc.js [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+
+  send (method, params) {
+    return this._processRequest(method, params)
+  }
+
+  register (method, handler) {
+    if (!this._handlerMap[method]) {
+      this._handlerMap[method] = handler
+    } else {
+      throw new Error(`Method ${method} already registered`)
```
Comment (rogaldh) [inline src/mqtt-rpc.js]: is already ...

File: src/mqtt-rpc.js [hunk 11]
```diff
@@ -0,0 +1,151 @@
+import { v4 as uuid4 } from 'uuid'
+
+class MQTTRPCService {
+  constructor (mqttClient, topicIn, topicOut, codec, methods) {
+    this._codec = codec
+    this._handlerMap = {}
```
Comment (rogaldh) [inline src/mqtt-rpc.js]: Why not to use Map instead?

File: src/service.js [hunk 12]
```diff
@@ -0,0 +1,82 @@
+import EventEmitter from 'events'
```
Comment (rogaldh) [inline src/service.js]: There may be problems with bundling as EventEmitter is not a browser API. So we may use EventTarget. All other envs may use polyfill for that.

File: src/service.js [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+
+    if (type === 'event' && payload !== undefined) {
+      event = {
+        type: label,
+        data: payload
+      }
+
+      this._ee.emit(event.type, event)
+    } else {
+      console.debug('Unknown message', topic, message)
```
Comment (rogaldh) [inline src/service.js]: console

File: src/service.js [hunk 14]
```diff
@@ -0,0 +1,82 @@
+import EventEmitter from 'events'
+
+import Codec from './codec'
+import MQTTRPCService from './mqtt-rpc'
+
+class Service {
+  constructor (mqttClient, agentId, appName) {
+    this._agentId = agentId
```
Comment (rogaldh) [inline src/service.js]: Validation

File: src/service.js [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+  }
+
+  destroy () {
+    this._mqtt.detachRoute(this._topicPatternNotifications)
+    this._ee.removeAllListeners()
+    this._rpc.destroy()
+  }
+}
+
+export default Service
```
Comment (rogaldh) [inline src/service.js]: Let's not use the default export

File: src/mqtt-rpc.js [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+  destroy () {
+    this._removeSubscription()
+
+    this._handlerMap = {}
+    this._incomingRequestMap = {}
+    this._requestStorage = {}
+  }
+}
+
+export default MQTTRPCService
```
Comment (rogaldh) [inline src/mqtt-rpc.js]: Let's not use the default export

File: src/event.js [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+      occurred_at: occurredAt,
+      room_id: roomId,
+      sets
+    }
+
+    return this._rpc.send('state.read', params)
+  }
+}
+
+export default Event
```
Comment (rogaldh) [inline src/event.js]: Let's not use the default export

File: src/conference.js [hunk 18]
```diff
@@ -0,0 +1,249 @@
...
+      room_id: roomId,
+      agent_id: agentId,
+      data
+    }
+
+    return this._rpc.send('message.unicast', params)
+  }
+}
+
+export default Conference
```
Comment (rogaldh) [inline src/conference.js]: Let's not use the default export

File: src/codec.js [hunk 19]
```diff
@@ -0,0 +1,16 @@
...
+  encode (data) {
+    return this._encoder(data)
+  }
+
+  decode (data) {
+    return this._decoder(data)
+  }
+}
+
+export default Codec
```
Comment (rogaldh) [inline src/codec.js]: Let's not use the default export

File: src/conference.js [hunk 20]
```diff
@@ -0,0 +1,249 @@
...
+
+  /**
+   * Create room
+   * @param {String} audience
+   * @param {[Number, Number]} time
+   * @param {String} backend
+   * @returns {Promise}
+   */
+  createRoom (audience, time, backend) {
+    const params = {
```
[the remark this answers is not in the document]
Comment (alexkonst) [reply src/conference.js]: I will make an issue for that.

File: src/event.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (alexkonst) [reply src/event.js]: Done

File: src/mqtt-rpc.js (hunk 3 above)
Comment (alexkonst) [reply src/mqtt-rpc.js]: Separate issue for that

File: src/mqtt-rpc.js (hunk 5 above)
Comment (alexkonst) [reply src/mqtt-rpc.js]: done

File: src/mqtt-rpc.js (hunk 6 above)
Comment (alexkonst) [reply src/mqtt-rpc.js]: fixed

File: src/mqtt-rpc.js (hunk 7 above)
Comment (alexkonst) [reply src/mqtt-rpc.js]: Separate issue for that

File: src/service.js (hunk 12 above)
Comment (alexkonst) [reply src/service.js]: Replaced with `events` npm package.

**Comment by alexkonst** [comment]
> Let's use eslint-config-adequate@2.0.2 preset. It extends the standard config and quite reliable.
I think that for small library `standard` preset will be enough.

**Comment by rogaldh** [comment]
Our custom config helps to prevent a lot of weird issues where `standard` does not help. Also, it's used widely at the company' projects.

**Pull request merged**